### PR TITLE
Extract interfaces for objects to be used through the executors widget

### PR DESCRIPTION
**Pull request by Vlatombe**
This is required for CloudBees CI HA support: we provide alternate implementations of these interfaces to represent computers and related objects that exist in different physical replicas of the same logical instance.

This allows to build an aggregated view of computers and executors, some local, some remote.

To avoid wording conflicts with existing concrete classes, the new interfaces are prefixed with `I` e.g. `IComputer`. I'm open to alternative naming suggestions, but unfortunately the names that would represent them the most accurately are already taken.

This introduces a new extension point, `ComputerSource`, allowing to override the list of computers that is returned via `ComputerSet`. Only the implementation with the highest ordinal among those registered will be picked up. The default implementation returns local `Computer` instances, and has ordinal `-1`, allowing a plugin to override it with a simple `@Extension` without any explicit ordinal being set.

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Computer.java:1737]: Moved to its own file.

File: core/src/main/java/hudson/model/Queue.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Queue.java:1976]: Moved to `ITask`

File: core/src/main/java/hudson/model/Computer.java
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:726]: I guess the `@see` could be pulled up as well.

File: core/src/main/java/hudson/model/ComputerSet.java
Comment (jglick) [inline core/src/main/java/hudson/model/ComputerSet.java:132]:
```suggestion
     * @since TODO
     */
```
or `Beta`

File: core/src/main/java/hudson/model/ComputerSet.java
Comment (jglick) [inline core/src/main/java/hudson/model/ComputerSet.java:530]: should be `DoNotUse`

File: core/src/main/java/jenkins/model/IComputer.java
Comment (jglick) [inline core/src/main/java/jenkins/model/IComputer.java:103]: I am not sure this annotation even works as documented. I recall a recent case where it did not. I would suggest just using `@CheckForNull`.

File: core/src/main/java/jenkins/model/IComputer.java
Comment (jglick) [inline core/src/main/java/jenkins/model/IComputer.java:101]: Empty string, or null? This is unclear.

File: core/src/main/java/jenkins/model/IComputer.java
Comment (jglick) [inline core/src/main/java/jenkins/model/IComputer.java:132]: There is no _default implementation_. If there is one in `Computer`, that should be noted there, with `@inheritDoc`.

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java:258]: Not binary-compatible (I think). Are there potentially any external callers?

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ITask.java:44]:
```suggestion
        if (this instanceof AccessControlled ac) {
            return ac.hasPermission(Item.CANCEL);
```

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ITask.java:55]:
```suggestion
        if (this instanceof AccessControlled ac) {
            return ac.hasPermission(Item.READ);
```

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ITask.java:70]: Unclear why one has a default and the other does not. What would a `null` return value mean?

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ITask.java:76]: Also undocumented what `null` would mean.

**Review by jglick — COMMENTED** [review]
Minor questions about Javadoc and nullability.

File: core/src/main/java/jenkins/model/IComputer.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/model/IComputer.java:103]: This is a mistake from my end, actually this is `@NonNull`.

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/model/queue/ITask.java]: Changed it to delegate to `getDisplayName`, which seems natural.

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java:270]: I checked but doesn't seem likely. I've retained the previous signature just in case.

File: core/src/main/java/hudson/model/Computer.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Computer.java:731]: Only here for `@Exported`

File: core/src/main/java/hudson/model/queue/SubTask.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/queue/SubTask.java:46]: While I'm here...

Comment (jglick) [reply core/src/main/java/hudson/model/queue/SubTask.java:46]: #2879 FTR

**Review by jglick — COMMENTED** [review]
Still unsure about `ITask.getUrl` (https://github.com/jenkinsci/jenkins/pull/9749#discussion_r1765062602).

File: core/src/main/java/jenkins/model/IComputer.java
Comment (jglick) [inline core/src/main/java/jenkins/model/IComputer.java:97]:
```suggestion
     * {@code hasOfflineCause() == true} implies this must be nonempty.
```
?

File: core/src/main/java/jenkins/model/IComputer.java
Comment (jglick) [inline core/src/main/java/jenkins/model/IComputer.java:90]: Now always true?

**Review by jglick — COMMENTED** [review]
https://github.com/jenkinsci/jenkins/pull/9749#discussion_r1765131546 still unclear

File: core/src/main/java/jenkins/model/IComputer.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/model/IComputer.java:90]: Ah right, this is inconsistent.

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/model/queue/ITask.java:76]: Looking into the possibility of leaving the method unimplemented.

Comment (Vlatombe) [reply core/src/main/java/jenkins/model/queue/ITask.java]: 4cf4a80

File: test/src/test/java/hudson/model/ProjectTest.java
Comment (jglick) [inline test/src/test/java/hudson/model/ProjectTest.java:958]: Maybe there should be a default impl in `SubTask` for compatibility reasons?

File: core/src/main/java/jenkins/model/queue/ITask.java
Comment (jglick) [inline core/src/main/java/jenkins/model/queue/ITask.java:74]: So this should now be `@CheckForNull`, given the default in `SubTask`?

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/model/ComputerSet.java
Comment (amuniz) [inline core/src/main/java/hudson/model/ComputerSet.java:129]: There might be a small change in the API with this, as `Collection`  does not grant ordering while the previous signature with an array was actually doing it. I don't think this will affect any client of this API, but just noting.

**Review by amuniz — APPROVED** [review]

File: core/src/main/java/hudson/model/ComputerSet.java
Comment (Vlatombe) [reply core/src/main/java/hudson/model/ComputerSet.java:129]: array doesn't grant it either, however indeed the sort call is missing from the new implementation. Will fix it.

**Review by amuniz — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after the pending security release, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/hudson/model/ComputerSet/index.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/ComputerSet/index.jelly:36]: In the wrong scope, will not be visible where it is needed. Seriously considering setting aside a week to prototype https://issues.jenkins.io/browse/JENKINS-31203 at this point!

**Review by jglick — CHANGES_REQUESTED** [review]

**Review by jglick — APPROVED** [review]

**Pull request merged**
